crates/compiler/src/compilation_steps/validate_unique_node_names.rs

```diff
@@ -0,0 +1,61 @@
+use crate::prelude::generated::yarnspinnerparser::{DialogueContextAttrs, NodeContextAttrs};
+use crate::prelude::*;
+use antlr_rust::token::Token;
+use std::collections::HashMap;
+
+pub(crate) fn validate_unique_node_names(
+    mut state: CompilationIntermediate,
+) -> CompilationIntermediate {
+    // Ensure that all nodes names in this compilation are unique. Node
+    // name uniqueness is important for several processes, so we do this
+    // check here.
+    let all_nodes = state.parsed_files.iter().flat_map(|file| {
+        file.tree
+            .node_all()
+            .iter()
+            .map(|node| (node.clone(), file))
+            .collect::<Vec<_>>()
+    });
+
+    // Pair up every node with its name, and filter out any that don't
+    // have a name
+    let nodes_with_names = all_nodes.filter_map(|(node, file)| {
+        node.header_all()
+            .iter()
+            .find(|header| header.header_key.as_ref().unwrap().get_text() == "title")
+            .map(|title_header| {
+                let title = title_header
+                    .header_value
+                    .as_ref()
+                    .unwrap()
+                    .get_text()
+                    .to_owned();
+                (title, title_header.clone(), file)
+            })
+    });
+
+    let nodes_by_name =
+        nodes_with_names.fold(HashMap::new(), |mut map, (name, header_context, file)| {
```

Comment:
can you create me an issue to refactor this? my functional heart is bleeding. this is way overcomplicated hahaha (not ur fault)